File: day18.c
#include <assert.h>
#include <limits.h>
#include <stdbool.h>
#include <stdio.h>
#include <string.h>

#define COUNT(a) (sizeof(a) / sizeof((a)[0]))

#define N (1 << 7)
#define A(x, y) a[(y) & (N - 1)][(x) & (N - 1)]
char a[N][N];
int n;

#define V(x, y) visit[(y) & (N - 1)][(x) & (N - 1)]
int visit[N][N];

const int dx[] = { 1, 0,-1, 0 };
const int dy[] = { 0, 1, 0,-1 };

typedef struct {
    int x, y;
} state_t;

state_t open[1 << 8];
int openrd, openwr;

void push(state_t s)
{
    open[openwr++] = s;
    openwr &= COUNT(open) - 1;
    assert(openrd != openwr);
}

state_t pop()
{
    assert(openrd != openwr);
    state_t s = open[openrd++];
    openrd &= COUNT(open) - 1;
    return s;
}

int bfs(state_t start, state_t end)
{
    for (int y = 0; y < n; y++) {
        for (int x = 0; x < n; x++) {
            V(x,  y) = INT_MAX;
        }
    }
    V(start.x, start.y) = 0;
    openrd = openwr = 0;
    push(start);

    while (openwr != openrd) {
        state_t s = pop();
        int dist = V(s.x, s.y);
        //A(s.x, s.y) = 'x';
        if (s.x == end.x && s.y == end.y) {
            return dist;
        }

        for (int d = 0; d < 4; d++) {
            state_t t = s;
            t.x += dx[d];
            t.y += dy[d];
            if (!A(t.x, t.y) || A(t.x, t.y) == '#') continue;
            int ndist = dist + 1;
            if (ndist < V(t.x, t.y)) {
                V(t.x, t.y) = ndist;
                push(t);
            }
        }
    }

    return INT_MAX;
}

int main()
{
    FILE* file = fopen("day18.txt", "r");
    //int limit = 12;
    //n = 6 + 1;
    int limit = 1024;
    n = 70 + 1;

    for (int y = 0; y < n; y++) {
        for (int x = 0; x < n; x++) {
            A(x, y) = '.';
        }
    }

    int cnt = 0;
    for (int x, y; 2 == fscanf(file, "%d,%d\n", &x, &y); ) {
        A(x, y) = '#';
        ++cnt;

        int dist = bfs((state_t){ 0, 0 }, (state_t){ n - 1, n - 1 });
        if (cnt == limit) {
            printf("%d\n", dist);
        }
        if (dist == INT_MAX) {
            printf("%d,%d\n", x, y);
            break;
        }
    }

    for (int y = 0; y < n; y++) {
        for (int x = 0; x < n; x++) {
            //printf("%c", A(x, y));
        }
        //printf("\n");
    }
}

File: day18.lua
#!/usr/bin/env lua

local a = {}
local n = 0

local dx = { 1, 0,-1, 0 }
local dy = { 0, 1, 0,-1 }

local function key(self)
    return self.x | self.y << 7
end

local function state(x, y)
    return { x = x, y = y, v = key }
end

local function push(self, v)
    local w = self.w or 1
    self[w] = v
    self.w = w + 1
end

local function pop(self)
    assert(self.r ~= self.w)
    local r = self.r or 1
    local v = self[r]
    self[r] = nil
    self.r = r + 1
    return v
end

local function empty(self)
    return self.r == self.w
end

local function bfs(start, ends)
    local visit = setmetatable({}, { __index = function() return math.maxinteger end })
    visit[start:v()] = 0
    local open = {}
    push(open, start)

    while not empty(open) do
        local s = pop(open)
        local dist = visit[s:v()]
        --a[s.y][s.x] = 'x'
        if s.x == ends.x and s.y == ends.y then
            return dist
        end

        for d = 1, 4 do
            local t = state(s.x, s.y)
            t.x = s.x + dx[d]
            t.y = s.y + dy[d]
            local ndist = dist + 1
            local c = a[t.y] and a[t.y][t.x]
            if c and c ~= '#' and ndist < visit[t:v()] then
                visit[t:v()] = ndist
                push(open, t)
            end
        end
    end
end

local file = io.open("day18.txt")
--local limit = 12
--n = 6
local limit = 1024
n = 70

for y = 0, n do
    a[y] = {}
    for x = 0, n do
        a[y][x] = '.'
    end
end

local coords = {}
for line in file:lines() do
    local xs, ys = line:match("(%d+),(%d+)")
    local x, y = tonumber(xs), tonumber(ys)
    a[y][x] = '#'
    coords[#coords + 1] = state(x, y)

    if #coords == limit then
        local dist = bfs(state(0, 0), state(n, n))
        print(dist)
    end
end

for i = #coords, 1, -1 do
    local s = coords[i]
    a[s.y][s.x] = '.'
    local dist = bfs(state(0, 0), state(n, n))
    if dist then
        print(s.x .. "," .. s.y)
        break
    end
end

for y = 0, n do
    for x = 0, n do
        --io.write(a[y][x])
    end
    --print()
end
